xks-axum/src/xks_proxy/pkcs11.rs

```diff
@@ -51,11 +56,87 @@ use crate::{settings, SETTINGS};
 
 const P11_CONTEXT_READ_FAILURE: &str = "Failed to acquire pkcs11 context";
 
+// https://stackoverflow.com/questions/69350600/how-to-reconcile-rust-mutex-and-c-caller-supplied-locking-mechanisms
+
+#[repr(C)]
+struct MutexContainer {
+    mutex: Mutex<()>,
+    guard: Cell<Option<MutexGuard<'static, ()>>>,
+}
+
+extern "C" fn create_mutex_container(ppcontainer: CK_VOID_PTR_PTR) -> CK_RV {
+    unsafe {
+        *ppcontainer = Box::into_raw(Box::new(MutexContainer {
+            mutex: Mutex::new(()),
+            guard: Cell::new(None),
+        }))
+        .cast();
+        tracing::trace!("create_mutex_container: *ppcontainer: {:?}", *ppcontainer);
+    }
+    CKR_OK
+}
+
+extern "C" fn destroy_mutex_container(pcontainer: CK_VOID_PTR) -> CK_RV {
+    tracing::trace!("destroy_mutex_container: pmutex_container: {pcontainer:?}");
+    unlock_mutex(pcontainer);
+    unsafe {
+        drop(Box::from_raw(pcontainer.cast::<MutexContainer>()));
+    }
+    CKR_OK
+}
+
+extern "C" fn lock_mutex(pcontainer: CK_VOID_PTR) -> CK_RV {
+    tracing::trace!("lock_mutex: pmutex_container: {pcontainer:?}");
+    let container: &MutexContainer = unsafe { &*pcontainer.cast() };
+    match container.mutex.lock() {
+        Ok(guard) => {
+            if let Some(existing) = container.guard.take() {
+                tracing::warn!("pmutex_container {pcontainer:?} already locked: existing={existing:?}, guard={guard:?}");
+            }
```

Comment:
`std::sync::Mutex::lock` does not return when trying to lock a mutex recursively. So if the guard were to be set at this point, it looks like a (potentially severe) bug and aborting with as much information as possible might be preferable, instead of continuing in a weird state with race conditions.

Reply:
Given `std::sync::Mutex::lock` does not return, `L93` will never get executed, Right?  So it would actually be a dead lock.

Perhaps I should change it to use `Mutex::try_lock` or replace `std::sync::Mutex::lock` with some re-entrant mutex if re-entrancy is possible.

Reply:
Just re-checked the [PKCS#11 spec](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html).  I should not change it to `try_lock` because

>  If a CK_LOCKMUTEX function is called on a mutex which is locked by some thread other than the calling thread, the calling thread blocks and waits for that mutex to be unlocked.

On the other hand,

> If a CK_LOCKMUTEX function is called on a mutex which is locked by the calling thread, the behavior of the function call is undefined.

So it seems changing to use a re-entrant lock is the only solution.

Reply:
Correct. Line 94 can only be executed if there's a bug somewhere else causing the Rust mutex to be unlocked but the guard to be set.

The pkcs11 spec states:
>  If a CK_LOCKMUTEX function is called on a mutex which is locked by the calling thread, the behavior of the function call is undefined

So it looks like you don't need to support recursively locking a mutex.

Reply:
I interpret that spec statement as: if this happens, this is a bug. Aborting is a safer alternative to continuing in a weird state where data races may happen.

Reply:
I agree, but in practice when such bug occurs, the mutex being locked twice from the same thread would dead lock.  In theory we don't need to handle that.  In practice, crashing is better than dead locking.

Reply:
> [...] crashing is better than dead locking

Agreed.

Reply:
I just tested the use of `parking_lot::ReentrantMutex`, and some weird things happened.  I saw the mutex being re-entrantly locked (which should not happen), and the application crashed.  This leads me to think that perhaps due to the use of async tokio runtime, the locking from SoftHSMv2 isn't thread-based?

Reply:
It might be simpler to just abort the process in the case where a guard is found after locking the mutex. Right now it logs, but continuing might lead to race conditions and corrupted data.

Reply:
Fixed in the latest revision.